### ISSUE TITLE
EvoCrash-generated tests vs user-written tests

### ISSUE THREAD

**Issue by vmassol** [STAMP-project/docs-forum#16]
Following up on the "[stamp] [D5.3] New KPIs after mid term review" thread, I'm posting here some analysis I did about EvoCrash. I would be curious to know what others think and if my reasoning is flawed.

For the context the discussion was about adding a KPI related to "Time to build a crash replicating test". However it also asks the question about how useful tests generated by EvoCrash are.

> As I hinted in my previous reply on this thread, there’s a big difference between:
> A) Understanding the stack trace
> B) Understanding the bug in the code and the real cause
>
> The “time it takes to reproduce a crash” is not a good industrial metric. A good metric would be “the time it takes to write a test showing the problem”. And that’s a lot harder.
>
> Here’s an example:
>
> I’m analyzing https://github.com/STAMP-project/evocrash-usecases-output/blob/master/XWIKI/XWIKI-13546/src/Xwiki_13546.java and even though EvoCrash was nicely able to reproduce the problem with 10 reproduced frames, it doesn’t help a little bit to fix the issue. Actually, finding out that the issue is that the wikiid is not set in the context is super easy (it takes 1 minute max for a dev to figure this out by looking at the last frame of the stack trace). However, understanding that the issue is that in the email preparation thread, we’re not storing the wikiid in the context and thus when sending the mail, it’s not present, is a lot harder and not something that EvoCrash is helping with. It also doesn’t help with writing the tests that needs to be written so that the problem doesn’t occur again. That tests is wildly different from the one generated by EvoCrash.

For the case of XWIKI-13546 (I did an analysis of it at https://github.com/STAMP-project/evocrash-usecases-output/tree/master/XWIKI/XWIKI-13546), the added value of EvoCrash is not that high IMO.

For me this raises the question of defining in which cases the tests generated by EvoCrash bring value over reading the stack trace.

Any idea?

**Comment by MaelAudren** [comment]
Indeed in the current version of EvoCrash, the generated tests are not relevant enough to be integrated in the source code. However I'm counting on the next upgrade of EvoCrash to provide tests that could be included as non regression tests.

**Comment by vmassol** [comment]
It would be interesting to have a characterization of the cases when evocrash-generated tests can be used as regression tests and when they cannot.

**Comment by xdevroey** [comment]
Yes, agreed. It would also be interesting to see what are the modifications done to transform a test generated by evocrash into a regression test.
